src/Math/include/BipedalLocomotion/Math/SchmittTrigger.h

```diff
@@ -0,0 +1,137 @@
+/**
+ * @file SchmittTrigger.h
+ * @authors Giulio Romualdi
+ * @copyright 2023 Istituto Italiano di Tecnologia (IIT). This software may be modified and
+ * distributed under the terms of the BSD-3-Clause license.
+ */
+
+#ifndef BIPEDAL_LOCOMOTION_MATH_SCHMITT_TRIGGER_H
+#define BIPEDAL_LOCOMOTION_MATH_SCHMITT_TRIGGER_H
+
+#include <memory>
+
+#include <BipedalLocomotion/ParametersHandler/IParametersHandler.h>
+#include <BipedalLocomotion/System/Advanceable.h>
+
+namespace BipedalLocomotion
+{
+namespace Math
+{
+
+/**
+ * SchmittTriggerState contains the internal state of the trigger,
+ */
+struct SchmittTriggerState
+{
+    bool state{false}; /**< current state*/
+    double switchTime{0.0}; /**< time instant at which the state was toggled in seconds */
+    double edgeTime{0.0}; /**< Time instant at which the raw value transited from low to high of
+                             from high to low in seconds.*/
+};
+
+/**
+ * SchmittTriggerState contains the input of the SchmittTrigger class.
+ */
+struct SchmittTriggerInput
+{
+    double time{0.0}; /**< Current time instant in seconds */
+    double rawValue{0.0} ; /**< Raw value that should  */
+};
+
+/**
+ * SchmittTrigger implements a discrete version of a SchmittTrigger
+ * See [here](https://en.wikipedia.org/wiki/Schmitt_trigger) for further details.
+ */
+class SchmittTrigger
+    : public BipedalLocomotion::System::Advanceable<SchmittTriggerInput, SchmittTriggerState>
+{
+public:
+    /**
+     * Struct holding switching parameters for the Schmitt Trigger
+     */
+    struct Params
+    {
+        double onThreshold{0.0}; /**< high value threshold to initiate an ON state switch after
+                                    switchOnAfter time-units*/
+        double offThreshold{0.0}; /**< low value threshold to initiate an OFF state switch after
+                                     switchOffAfter time-units*/
+        double switchOnAfter{0.0}; /**< time units to wait for before switching to ON state from OFF
+                                      state. Ensure it's greater than sampling time. */
+        double switchOffAfter{0.0}; /**< time units to wait for before switching to OFF state from
+                                       ON state. Ensure it's greater than sampling time. */
+
+        /** Threshold used for the comparison of two time instant. Given two time instants if the
+         * error between the two is lower than the threshold, the time instants are considered
+         * equal. */
+        double timeComparisonThreshold{std::numeric_limits<double>::epsilon()};
```

Comment:
Why ``epsilon`` and not ``min``?

Reply:
You are right. At this point, I think that `epsilon` is better than min indeed as explained [here](https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon)

- **`std::numeric_limits<double>::epsilon()`** Returns the machine epsilon, that is, the difference between 1.0 and the next value representable by the floating-point type T.

- **`std::numeric_limits<double>::min()`** Returns the minimum finite value representable by the numeric type T.




In https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon, shows how to use `std::numeric_limits<double>::epsilon()` to compare floating-point values for equality.

Reply:
We may implement something like this in `Math` component and use it everywhere is needed 
```c++
template<class T>
typename [std::enable_if](http://en.cppreference.com/w/cpp/types/enable_if)<![std::numeric_limits](http://en.cppreference.com/w/cpp/types/numeric_limits)<T>::is_integer, bool>::type
    almost_equal(T x, T y, int ulp)
{
    // the machine epsilon has to be scaled to the magnitude of the values used
    // and multiplied by the desired precision in ULPs (units in the last place)
    return [std::fabs](http://en.cppreference.com/w/cpp/numeric/math/fabs)(x - y) <= [std::numeric_limits](http://en.cppreference.com/w/cpp/types/numeric_limits)<T>::epsilon() * [std::fabs](http://en.cppreference.com/w/cpp/numeric/math/fabs)(x + y) * ulp
        // unless the result is subnormal
        || [std::fabs](http://en.cppreference.com/w/cpp/numeric/math/fabs)(x - y) < [std::numeric_limits](http://en.cppreference.com/w/cpp/types/numeric_limits)<T>::min();
}
```

Reply:
That's where my doubt arose. ``epsilon`` seems to work better for checking if you are close to 1, while in https://github.com/ami-iit/bipedal-locomotion-framework/blob/f3241a5471b44e5f2815a32c06c9c82aeca51829/src/Math/src/SchmittTrigger.cpp#L77 you are checking a difference that should be close to zero. Note that the example of https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon have another part on the right hand side. In order to copy it you should do
```cpp
(std::abs(a - b) <= this->m_params.timeComparisonThreshold) * std::abs(a + b) )
```
but I don't think it is worth adding the additional multiplication. In addition, it makes the check more and more coarse the higher is ``a+b``. That's why they also added the 
```cpp
|| std::fabs(x - y) < std::numeric_limits<T>::min();
```
in the check (again using ``min``).

In any case, we are talking about an infinitesimal difference, so feel free to use the version you prefer :blush:

Reply:
> We may implement something like this in `Math` component and use it everywhere is needed (Taken from https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon)
> 
> ```c++
> template<class T>
> typename std::enable_if<!std::numeric_limits<T>::is_integer, bool>::type
>     almost_equal(T x, T y, int ulp)
> {
>     // the machine epsilon has to be scaled to the magnitude of the values used
>     // and multiplied by the desired precision in ULPs (units in the last place)
>     return std::fabs(x - y) <= std::numeric_limits<T>::epsilon() * std::fabs(x + y) * ulp
>         // unless the result is subnormal
>         || std::fabs(x - y) < std::numeric_limits<T>::min();
> }
> ```

Loving it, but I guess it is a bit of an over-complexification :thinking:

Reply:
Here I am! So I tried to do some tests following this [nice post](https://embeddeduse.com/2019/08/26/qt-compare-two-floats/) with a toy problem implemented in [floating-point-comparison](https://github.com/GiulioRomualdi/floating-point-comparison). Here I implemented for type of comparisons
```c++

template <typename T>
bool comparison0(T f1, T f2, T tolerance)
{
    return f1 == f2;
}

template <typename T>
bool comparison1(T f1, T f2, T tolerance)
{
    return std::abs(f1 - f2) <= tolerance;
}

template <typename T>
bool comparison2(T f1, T f2, T tolerance)
{
    return std::abs(f1 - f2) <= tolerance * std::max(std::abs(f1), std::abs(f2));
}

template <typename T>
bool comparison3(T f1, T f2, T tolerance)
{
    std::size_t ulp = 2;

    return std::fabs(f1 - f2) <= tolerance * std::fabs(f1 + f2) * ulp
        || std::fabs(f1 - f2) < std::numeric_limits<T>::min();
}
```
The equalities are tested for  from `9.0` to `99` in steps of `0.01` 

Setting the tolerance equal to **`1e-10`** We got the following results where **correct** means the number of comparisons that were correct 
| Comparison Function | correct | total |
|:---:|:--:|:---:|
| 0 | 10 | 10000 |
| 1 | 10000 | 10000 |
| 2 | 9999 | 10000 |
| 3 | 9999 | 10000 |

On the other hand setting the tolerance equal to **`std::numeric_limit<double>::epsilon()`**  I got the following

| Comparison Function | correct | total |
|:---:|:--:|:---:|
| 0 | 24 | 10000 |
| 1 | 42 | 10000 |
| 2 | 70 | 10000 |
| 3 | 274 | 10000 |

The following conclusion holds (not in general but in our case):
- **`std::numeric_limit<double>::epsilon()`** is too low tolerance
- even if [here](https://embeddeduse.com/2019/08/26/qt-compare-two-floats/), and [here](https://stackoverflow.com/questions/67842444/modern-practice-to-compare-double-float-for-equality-in-modern-c) suggested to avoid considering an absolute threshold for us `comparison1()` seems to be enough.

So we should decide the tolerance we want to guarantee in blf. (We can store it as parameter in Math component along with the Gravity)